javaslang/src/main/java/javaslang/collection/PriorityQueue.java

```diff
@@ -0,0 +1,791 @@
+
+        PriorityQueue<T> result = empty(comparator);
+        for (T t : this) {
+            if (predicate.test(t)) {
+                break;
+            }
+            result = result.insert(t);
+        }
+        return result;
+    }
+
+    @Override
+    public <T1, T2> Tuple2<? extends PriorityQueue<T1>, ? extends PriorityQueue<T2>> unzip(Function<? super T, Tuple2<? extends T1, ? extends T2>> unzipper) {
+        Objects.requireNonNull(unzipper, "unzipper is null");
+        final Tuple2<Iterator<T1>, Iterator<T2>> unzip = iterator().unzip(unzipper);
+        return Tuple.of(ofAll(naturalComparator(), unzip._1), ofAll(naturalComparator(), unzip._2));
+    }
+
+    @Override
+    public <T1, T2, T3> Tuple3<? extends PriorityQueue<T1>, ? extends PriorityQueue<T2>, ? extends PriorityQueue<T3>> unzip3(Function<? super T, Tuple3<? extends T1, ? extends T2, ? extends T3>> unzipper) {
+        Objects.requireNonNull(unzipper, "unzipper is null");
+        final Tuple3<Iterator<T1>, Iterator<T2>, Iterator<T3>> unzip3 = iterator().unzip3(unzipper);
+        return Tuple.of(ofAll(naturalComparator(), unzip3._1), ofAll(naturalComparator(), unzip3._2), ofAll(naturalComparator(), unzip3._3));
+    }
+
+
+    @Override
+    public <U> PriorityQueue<Tuple2<T, U>> zip(Iterable<? extends U> that) {
+        Objects.requireNonNull(that, "that is null");
+        return ofAll(iterator().zip(that));
+    }
+
+    @Override
+    public <U> PriorityQueue<Tuple2<T, U>> zipAll(Iterable<? extends U> that, T thisElem, U thatElem) {
+        Objects.requireNonNull(that, "that is null");
+        return ofAll(iterator().zipAll(that, thisElem, thatElem));
+    }
+
+    @Override
+    public PriorityQueue<Tuple2<T, Long>> zipWithIndex() {
+        return ofAll(iterator().zipWithIndex());
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (o == this) {
+            return true;
+        } else if (o instanceof PriorityQueue) {
+            final PriorityQueue<?> that = (PriorityQueue<?>) o;
+            return this.toList().equals(that.toList());
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        return toList().hashCode();
+    }
+
+    @Override
+    public String stringPrefix() {
+        return "PriorityQueue";
+    }
+
+    @Override
+    public String toString() {
+        return mkString(stringPrefix() + "(", ", ", ")");
+    }
+
+    public PriorityQueue<T> padTo(int length, T element) {
+        final int actualLength = length();
+        if (length <= actualLength) {
+            return this;
+        } else {
+            return ofAll(comparator, toList().padTo(length, element));
+        }
+    }
+}
+
+abstract class PriorityQueueBase<T, Q extends PriorityQueueBase<T, Q>> extends AbstractsQueue<T, Q> {
+    protected final SerializableComparator<T> comparator;
+    protected final List<Node<Ranked<T>>> forest;
```

Comment:
What is the rationale behind storing a list of trees? (I know the answer is in the code, I just haven't had the time to dig too deeply). Does each element in the list has a tree of nodes with same rank?